### ISSUE TITLE
ingress-nginx doesn't receive updated configuration on cluster configuration change

### ISSUE THREAD

**Issue by ChromoX** [rancher/rancher#15079]
**Rancher versions:**
rancher/server or rancher/rancher: 2.0.7
rancher/agent or rancher/rancher-agent: 2.0.7

**Infrastructure Stack versions:**
kubernetes (if applicable): 1.10.5

**Docker version: (`docker version`,`docker info` preferred)**

> Server:
 Engine:
  Version:      18.03.1-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.5
  Git commit:   9ee9f40
  Built:        Thu Apr 26 07:23:03 2018
  OS/Arch:      linux/amd64
  Experimental: false

**Operating system and kernel: (`cat /etc/os-release`, `uname -r` preferred)**
RancherOS 1.4.0: 4.14.32-rancher2

**Type/provider of hosts: (VirtualBox/Bare-metal/AWS/GCE/DO)**
Bare-metal 

**Setup details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
Single node rancher, internal db.

**Environment Template: (Cattle/Kubernetes/Swarm/Mesos)**

**Steps to Reproduce:**
1. Setup cluster with default settings.
2. Modify cluster configuration(editing YAML) to include a few extra/different fields.
    * This first update will propagate and work.
3. Modify cluster configuration again to include some changes to ingress-nginx settings.

**Results:**
New ingress-nginx settings are not pushed to nodes.

**Comment by spatialy** [comment]
Hi @ChromoX, same problem here.

any solutions???

best

**Comment by ChromoX** [comment]
@spatialy The only solution I've found thus far is to remake the cluster.

**Comment by alena1108** [comment]
@galal-hussein we'll have to handle the addon upgrades on the rke side

**Comment by galal-hussein** [comment]
@alena1108 we already handle addon upgrade since 2.0.7
@ChromoX i tested using version 2.0.7 first trying to pass the option:
```
ssl-protocols: SSLv2
```
to ingress, and i was able to confirm that the option is correct in the configmap, afterwards i edited the cluster.yml and added another option:
```
ssl-protocols: "SSLv2"
proxy-buffer-size: "4k"
```
i can see that rke upgraded and nginx config map upgraded correctly:
```
ubectl get configmap/nginx-configuration -o yaml -n ingress-nginx
apiVersion: v1
data:
  proxy-buffer-size: 4k
  ssl-protocols: SSLv2
kind: ConfigMap
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","data":{"proxy-buffer-size":"4k","ssl-protocols":"SSLv2"},"kind":"ConfigMap","metadata":{"annotations":{},"labels":{"app":"ingress-nginx"},"name":"nginx-configuration","namespace":"ingress-nginx"}}
.....
```
i was able to see also the following in the ingress nginx controller pod:
```
> kubectl logs -n ingress-nginx nginx-ingress-controller-prgf9
...............
I0828 22:09:36.012664       7 controller.go:169] Configuration changes detected, backend reload required.
I0828 22:09:36.217786       7 controller.go:179] Backend successfully reloaded.
I0828 22:31:47.033373       7 event.go:218] Event(v1.ObjectReference{Kind:"ConfigMap", Namespace:"ingress-nginx", Name:"nginx-configuration", UID:"0ca93f9f-ab0f-11e8-8582-0657b92e9eb6", APIVersion:"v1", ResourceVersion:"3432", FieldPath:""}): type: 'Normal' reason: 'UPDATE' ConfigMap ingress-nginx/nginx-configuration

**Comment by ChromoX** [comment]
```yaml
ingress: 
  extra_args: 
    enable-ssl-passthrough: "true"
    tcp-services-configmap: "ingress-nginx/tcp-services"
  provider: "nginx"
```

My ingress settings are above. I added the ```tcp-services-configmap``` and it is not pushed it seems.

**Comment by galal-hussein** [comment]
@ChromoX i tried your configuration and it worked as expected, i can see the extra args you added appended to the nginx controller command line:
```
kubectl get daemonset -n ingress-nginx -o yaml
....
        - args:
          - /nginx-ingress-controller
          - --default-backend-service=$(POD_NAMESPACE)/default-http-backend
          - --configmap=$(POD_NAMESPACE)/nginx-configuration
          - --tcp-services-configmap=$(POD_NAMESPACE)/tcp-services
          - --udp-services-configmap=$(POD_NAMESPACE)/udp-services
          - --annotations-prefix=nginx.ingress.kubernetes.io
          - --enable-ssl-passthrough=true
          - --tcp-services-configmap=ingress-nginx/tcp-services
....
```

**Comment by ChromoX** [comment]
As I said above I have since remade the cluster and therefore I can't test exactly on the old cluster.

One slight state differece I remember my old cluster having is that at first I had set ```tcp-services-configmap``` to ```tcp-services```, which caused an error. Then I attempting updating to fix the error and that's when it wouldn't work.

**Comment by galal-hussein** [comment]
@ChromoX thanks for the info, creating the configmap makes sense since the controller needs it, however it nginx doesn't seem to exit without it. updates are available since 2.0.7 so any updates to nginx controller in the yaml will be reflected.

I will close the issue since i can't reproduce, please feel free to re-open if you can reproduce it and if so please check the job deployment logs for nginx controller deployment.
